### ISSUE TITLE
hstrerror() linking error when compile on HP-UX

### ISSUE THREAD

**Issue by pwpo1999** [openssl/openssl#2091]
I built libcrypto and libssl.
After, I attempt to link libraries into my app.
But, An error occurred with this message "Unsatisfied Symbol hstrerrer".

I found that function. src/crypto/bio/b_addr.c
Also, I found that hstrerror() is a function not supported on HP-UX.

Please confirm. Thank you.

**Comment by richsalz** [comment]
Try adding -lresol to your link line.

**Comment by pwpo1999** [comment]
I've done your advice, but I can not.
hstrerror() is missing on HP-UX 11
https://www.gnu.org/software/gnulib/manual/html_node/hstrerror.html

**Comment by richsalz** [comment]
Show the commandline how you linked your program, and the output.  (It should be -lresolv BTW)

**Comment by pwpo1999** [comment]
gcc -g -o main.o ../mylib/libmylib.a ../openssl/lib/libcrypto.a -lresolv
output - /usr/ccs/bin/ld: Can't find library: "resolv"
libresolv not exist...

**Comment by levitte** [comment]
@richsalz, there are numerous reports that `hstrerror()` is indeed missing on HP-UX.  No `-lresolv` will help.

Hold on, I'll see if I can fix something up.

**Comment by levitte** [comment]
@pwpo1999, are you trying to build the master branch or the 1.1.0 branch?

**Comment by pwpo1999** [comment]
I built Release version 1.1.0c

**Comment by levitte** [comment]
Ok cool.  Would you please try the change in #2092 to see if that solves your problem?

**Comment by pwpo1999** [comment]
I tried #2092. Success with no problem

**Comment by levitte** [comment]
Cool, thanks

**Comment by dot-asm** [comment]
Under the circumstances I'd say why not *remove* hstrerror call altogether? The code actually goes like this:
```
    if (1) {
        int gai_ret = 0;
#ifdef AI_PASSIVE
        ...
    } else {
#endif
        ... hstrerror()
    }
```
It should be noted that on all contemporary OSes, HP-UX included!, the `else` clause is not actually executed. And optimizing compilers would routinely remove that call to hstrerror potentially avoiding the linking problem. But one can also question legitimacy of using hstrerror. For example on Solaris, because it's not directly associated with gethostbyname, but with the BIND API. Even on HP-UX gethostbyname manual page says "If the name server is not being used, the value of h_errno may not be meaningful." In other words if /etc/hosts (or maybe even NIS?) is all you have, then h_errno is meaningless.

To summarize. It's never actually called on contemporary systems. Even if it was, it's not given that it would be meaningful in all cases without exclusion. And then question would be why not just avoid the call and pass "system" error for application to deal with as developer considers fit? I mean just SYSerr it...

And while we're at it. It would be appropriate to move `int gai_ret` inside #ifdef AI_PASSIVE. And one can also remove -lresolv from Solaris targets...

**Comment by richsalz** [comment]
I didn't think of things that way.  I think @dot-asm is right.

**Comment by dot-asm** [comment]
BTW, there are two calls to hstrerror, one after getservbyname. But I don't see that h_errno/hstrerror applies to it...

**Comment by dot-asm** [comment]
I mean hstrerror after getservbyname appears unjustified...

**Comment by levitte** [comment]
> I mean hstrerror after getservbyname appears unjustified...

Probably my misunderstanding...  I thought `getservbyname` reported errors in the same manner as `gethostbyname`...

**Comment by dot-asm** [comment]
Just to clarify, I'm not suggesting to remove the `else` clause (it should still be useful on legacy systems), only scrap calls to hstrerror...

**Comment by levitte** [comment]
> Even on HP-UX gethostbyname manual page says "If the name server is not being used, the value of h_errno may not be meaningful." In other words if /etc/hosts (or maybe even NIS?) is all you have, then h_errno is meaningless.

How *utterly* confusing.  How does one know when it's meaningful and when it's not?

**Comment by dot-asm** [comment]
> How does one know when it's meaningful and when it's not?

You don't, but system administrator should be able to make judgement if presented with "syserr"-ed code. But once again, it's actually meaningless question in contemporary OS context, because they will use if(1) thing with gai_strerror. Well, there is contradiction in what I'm saying. In sense that assertion that it's meaningless in contemporary OS context doesn't mean that it's meaningless in legacy OS context. But then question is what are these legacy OSes? Even if you can name one, I'd still say that it's not unreasonable compromise to "punish" it a little bit in name of better portability(`*`). Note that we are not talking about execution flow, but how error conditions are presented to application.

(`*`) Note that hstrerror is claimed obsolete, so that there is no guarantee that it won't start disappearing here and there...

**Comment by mattcaswell** [comment]
This issue was fixed by #2092
